### PR TITLE
feat(optimizer): implement 2-phase stream agg

### PR DESCRIPTION
**Pull request by kwannoel**
## What's changed and what's your intention?

As per title:
```
LocalAgg (single-value) -> Exchange -> GlobalAgg
```

- Summarize your change (**mandatory**)
  Perform 2-phase agg if input is distributed.

- How does this PR work? Need a brief introduction for the changed logic (optional)
  When converting from logical plan to stream plan, if input is distributed, insert a `LocalAgg` and `Exchange(single)` to do 2-phase aggregation.

- Describe clearly one logical change and avoid lazy messages (optional)
  A new plan node, `LocalSimpleAgg` is included, since the semantics of `LocalSimpleAgg` differs from `GlobalSimpleAgg`.
  It is used for distributed input, does not persist state. Additionally output is append-only, compared to `GlobalSimpleAgg` which has stateful output.

- Describe any limitations of the current code (optional)
  Currently does not support `non-append-only min/max`.
  In the future perhaps `LocalSimpleAgg` should be renamed to `LocalAppendOnlySimpleAgg`. It is an optimization on top of the more general stateful `LocalSimpleAgg`.
  We will also need to replace the single value `LocalSimpleAgg` with the stateful one, since it is more general.

- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional)
  NIL

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

https://github.com/singularity-data/risingwave/issues/2997

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 1]
```diff
@@ -281,3 +304,32 @@
...
+              StreamHashAgg { group_keys: [$0], aggs: [count, sum($1)] }
+                StreamExchange { dist: HashShard([0]) }
+                  StreamTableScan { table: t, columns: [v1, v2, _row_id], pk_indices: [2] }
+- sql: |
+    /* simple-stream-agg with append-only min-max should use 2-phase-agg */
+    create table t(v1 int, v2 int) with ('appendonly' = true);
+    select min(v1) as a1, max(v2) as a2 from t;
+  stream_plan: |
+    StreamMaterialize { columns: [agg#0(hidden), a1, a2], pk_columns: [] }
+      StreamSimpleAgg { aggs: [sum($0), min($1), max($2)] }
```
Comment (kwannoel) [inline src/frontend/test_runner/tests/testdata/agg.yaml:322]: This should be append only too.

Comment (kwannoel) [reply src/frontend/test_runner/tests/testdata/agg.yaml:320]: Actually seems like we can't support 2-phase min-max at all?

Even if input to partial agg step is append-only, output stream may not be append-only.

**Comment by kwannoel** [comment]
Currently failing:
```
dev=> create table t1 (v1 int not null, v2 int not null, v3 int not null);
CREATE_TABLE
dev=> 
dev=> insert into t1 values (1,4,2), (2,3,3);
INSERT 0 2
dev=> 
dev=> explain create materialized view mv2 as select round(avg(v1), 1) as avg_v1, sum(v2) as sum_v2, count(v3) as count_v3 from t1;
                                            QUERY PLAN                                             
---------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [avg_v1, sum_v2, count_v3], pk_columns: [] }
   StreamProject { exprs: [RoundDigit(($1::Decimal / $2), 1:Int32), $3, $4] }
     StreamSimpleAgg { aggs: [sum($0), sum($1), sum($2), sum($3), sum($4)], pk_indices: [] }
       StreamExchange { dist: Single, pk_indices: [] }
         StreamSimpleAgg { aggs: [count, sum($0), count($0), sum($1), count($2)], pk_indices: [] }
           StreamTableScan { table: t1, columns: [v1, v2, v3, _row_id], pk_indices: [3] }
(6 rows)

dev=> create materialized view mv2 as select round(avg(v1), 1) as avg_v1, sum(v2) as sum_v2, count(v3) as count_v3 from t1;ERROR:  RPC error: GrpcStatus(Status { code: Unknown, message: "transport error", source: Some(tonic::transport::Error(Transport, hyper::Error(Io, Kind(BrokenPipe)))) })
```

Fails as it tries to serialize `pk: []` with some order types. Need to write some integration tests to figure out what's going on

**Comment by skyzh** [comment]
I guess the issue is caused by wrong induction of state table schema. cc @BowenXiao1999 any ideas?

**Comment by skyzh** [comment]
... might be related to this part:

https://github.com/singularity-data/risingwave/blob/309ce36470483c6240d644936e916b5521178d6c/src/stream/src/executor/aggregation/mod.rs#L400-L429

**Comment by BowenXiao1999** [comment]
Let me take a look

**Comment by BowenXiao1999** [comment]
https://github.com/singularity-data/risingwave/blob/dc8631e6bbf377afcdac0bffda070768d6e08bb1/src/stream/src/executor/managed_state/aggregation/value.rs#L53-L57

For value states (except min, max), if there is no group key, the pk must be empty.

**Comment by kwannoel** [comment]
Closing this in favour of stateful approach to be used: https://singularity-data.quip.com/KtaRA6CspqRK/RFC-2-Phase-Agg-TopN-Operator-in-Streaming. It will be able to handle non-append only min / max too.

**Comment by skyzh** [comment]
I think we'll still need this. The stateless approach can only be used on append-only min max and value state, and the stateful approach can only be used on max min topN.

File: src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+impl StreamLocalSimpleAgg {
+    pub fn new(logical: LogicalAgg) -> Self {
+        let ctx = logical.base.ctx.clone();
+        let pk_indices = logical.base.pk_indices.to_vec();
+        let input = logical.input();
+        let input_dist = input.distribution();
+        debug_assert!(input_dist.satisfies(&RequiredDist::AnyShard));
+
+        let append_only = input.append_only();
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs:39]: `AggCall` uses plan node's `append_only` field: https://github.com/singularity-data/risingwave/blob/91a699062b6ea2d20f4f81b485f8334b283b36d8/src/stream/src/from_proto/local_simple_agg.rs#L37-L53.

`append_only` of GlobalAgg depends on its inputs. Which can be traced back to `stream_local_simple_agg`.

We need `AggCall` to be `append_only` to use `SingleValue` state rather than extreme value:

https://github.com/singularity-data/risingwave/blob/91a699062b6ea2d20f4f81b485f8334b283b36d8/src/stream/src/executor/managed_state/aggregation/mod.rs#L129-L134

File: src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        self.logical.agg_calls()
+    }
+}
+
+impl fmt::Display for StreamLocalSimpleAgg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut builder = if self.input().append_only() {
+            f.debug_struct("StreamAppendOnlyLocalSimpleAgg")
+        } else {
+            f.debug_struct("StreamLocalSimpleAgg")
```
Comment (kwannoel) [inline src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs:63]: It should always be append only.

File: src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        let ctx = logical.base.ctx.clone();
+        let pk_indices = logical.base.pk_indices.to_vec();
+        let input = logical.input();
+        let input_dist = input.distribution();
+        debug_assert!(input_dist.satisfies(&RequiredDist::AnyShard));
+
+        // Single-value local simple agg only produces inserts
+        // it should be append-only, even if input is non-append.
+        // N.B. This assumes input is valid - non-append-only min/max is rejected.
+        let append_only = true;
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs:42]: The semantics of local simple agg is somehow strange (it cannot be materialized). So I'd prefer to set append_only false?

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 5]
```diff
@@ -701,21 +701,67 @@ impl ToBatch for LogicalAgg {
...
         if self.group_keys().is_empty() {
-            Ok(StreamSimpleAgg::new(
-                self.clone_with_input(
-                    self.input()
-                        .to_stream_with_dist_required(&RequiredDist::single())?,
-                ),
-            )
-            .into())
+            let mut has_min_or_max_agg_calls = false;
+            for c in &self.agg_calls {
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:708]:
```
self.agg_calls.iter().any(|c| matches!(c.agg_kind, AggKind::Min | AggKind::Max))
```

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 6]
```diff
@@ -701,21 +701,67 @@ impl ToBatch for LogicalAgg {
...
+            let input_stream = input.to_stream()?;
+            let is_append_only = input_stream.append_only();
+            let input_distribution = input_stream.distribution();
+
+            // simple 2-phase-agg
+            if input_distribution.satisfies(&RequiredDist::AnyShard)
+                // TODO: non-append-only min/max is not supported yet,
+                // since min/max is a table rather than single value for local agg.
+                // see: https://github.com/singularity-data/risingwave/issues/2997#issuecomment-1156139576
+                && (is_append_only || !has_min_or_max_agg_calls)
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:727]: I'd prefer to have a standalone `allow_two_phase_agg` that takes all agg calls and `append_only` as parameter. In the future we will have more and more agg calls like `string_agg`, `count_distinct`, etc., and I'm not sure whether all of them can do two-phase. So having a single function for this can make it clearer for developers to know what is supported for now.

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 7]
```diff
@@ -701,21 +701,67 @@ impl ToBatch for LogicalAgg {
...
+            {
+                // partial agg
+                let partial_agg_plan =
+                    StreamLocalSimpleAgg::new(self.clone_with_input(input_stream)).into();
+
+                // insert exchange
+                let exchange = StreamExchange::new(partial_agg_plan, Distribution::Single).into();
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:734]: You don't need to insert exchange manually. As long as you set global agg to `Single` distribution, there will be a phase to insert exchange afterwards IIRC.

Comment (skyzh) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs:734]: In some places `enforce_if_not_satisfies` will be called on stream plan nodes, you can have a try not manually inserting the exchange.

**Review by skyzh — COMMENTED** [review]
Rest LGTM, good work!

File: src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs (hunk 4 above)
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs:42]: Hmm I used it for this purpose: https://github.com/singularity-data/risingwave/pull/3320#discussion_r906204180.

TLDR: global agg uses single value state if input is `append_only`. Otherwise it may use `extreme` agg state e.g. for `min/max`. That means we can't support append_only min/max.

Perhaps we can change `append_only` to `append_only_output`? I'll try to think of some alternatives.

Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/stream_local_simple_agg.rs:42]: We can defer `append_only` min/max for now I guess. Since stateful approach will deal with it.

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 8]
```diff
@@ -148,10 +148,11 @@
           BatchScan { table: t, columns: [v1, v2] }
   stream_plan: |
     StreamMaterialize { columns: [agg#0(hidden), cnt, sum], pk_columns: [] }
-      StreamSimpleAgg { aggs: [count, count($0), sum($0)] }
+      StreamAppendOnlySimpleAgg { aggs: [sum($0), sum($1), sum($2)] }
```
Comment (BowenXiao1999) [inline src/frontend/test_runner/tests/testdata/agg.yaml:151]: Why we always call the GlobalAgg as `AppendOnly`?

**Review by BowenXiao1999 — COMMENTED** [review]
Rest LGTM, Good work!

File: src/frontend/src/optimizer/plan_node/logical_agg.rs (hunk 7 above)
Comment (kwannoel) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs]: Okay I call `enforce_if_not_satisfies` for more declarative approach.

File: src/frontend/test_runner/tests/testdata/agg.yaml (hunk 8 above)
Comment (kwannoel) [reply src/frontend/test_runner/tests/testdata/agg.yaml]: Previously that was because I assumed local agg's to always be append_only. Have updated it.

See discussion: https://github.com/singularity-data/risingwave/pull/3320#discussion_r906628619.

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 9]
```diff
@@ -701,21 +726,39 @@ impl ToBatch for LogicalAgg {
...
+        // simple-agg
         if self.group_keys().is_empty() {
-            Ok(StreamSimpleAgg::new(
-                self.clone_with_input(
-                    self.input()
-                        .to_stream_with_dist_required(&RequiredDist::single())?,
-                ),
-            )
-            .into())
+            let has_min_or_max_agg_calls = self
```
Comment (skyzh) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:732]: I'd prefer to explicitly write supported two-phase aggs instead of writing unsupported one, because we will never know whether newly-added aggs can be two-phase.

Maybe:

```
let use_two_phase_agg = xxxx.iter().all(|c| matches!(c.agg_kind, AggKind::Count | AggKind::Sum)));
```

**Review by skyzh — APPROVED** [review]
LGTM, good work!
